bundles/org.eclipse.emfcloud.modelserver.common/src/org/eclipse/emfcloud/modelserver/common/patch/AbstractJsonPatchHelper.java

```diff
@@ -0,0 +1,716 @@
+            // This is actually a move. First, remove the Object from its original parent.
+            // This is not required to actually apply the change (EMF will do it automatically),
+            // however this is necessary to properly undo the change.
+            result = RemoveCommand.create(getEditingDomain(objectToAdd), objectToAdd.eContainer(),
+               objectToAdd.eContainingFeature(),
+               objectToAdd);
+         }
+      }
+
+      Command create = AddCommand.create(getEditingDomain(setting.getEObject()), setting.getEObject(),
+         setting.getFeature(), Collections.singleton(objectToAdd));
+      result = result == null ? create : result.chain(create);
+      return result;
+   }
+
+   protected final ResourceSet getResourceSet(final EObject objectToAdd) {
+      return objectToAdd == null || objectToAdd.eResource() == null ? null : objectToAdd.eResource().getResourceSet();
+   }
+
+   @SuppressWarnings("checkstyle:CyclomaticComplexity")
+   protected EObject getObjectToAdd(final String modelURI, final ResourceSet resourceSet, final JsonNode value)
+      throws JsonPatchException {
+      JsonNode idAttr = value.get(JsonConstants.ID_ATTR);
+      if (idAttr != null) {
+         // If the ID is specified, we're adding an existing object to a list (or a different parent).
+         // Find the existing object (or return null if the $id is invalid).
+         URI eObjectURI = URI.createURI(idAttr.asText());
+         return getEObject(modelURI, eObjectURI);
+      }
+      // Otherwise, create a new Object of the specified $type (or return null if the $type is invalid)
+      String type = value.get(JsonConstants.TYPE_ATTR).asText();
+      if (type == null) {
+         return null;
+      }
+      EClassifier eClassifier = getEClass(modelURI, resourceSet, type);
+      if (eClassifier == null) {
+         return null;
+      }
+
+      if (eClassifier instanceof EClass) {
+         EClass eClass = (EClass) eClassifier;
+         EFactory eFactoryInstance = eClass.getEPackage().getEFactoryInstance();
+         EObject newObject = eFactoryInstance.create(eClass);
+         Iterator<Entry<String, JsonNode>> fields = value.fields();
+         while (fields.hasNext()) {
+            Entry<String, JsonNode> field = fields.next();
+            if (field.getKey().startsWith(JsonConstants.METADATA_PREFIX)) {
+               continue;
+            }
+            EStructuralFeature feature = eClass.getEStructuralFeature(field.getKey());
+            if (feature == null) {
+               LOG.warn("Ignored unknown field: " + field.getKey());
+               continue;
+            }
+
+            Object emfValue = getEMFValue(modelURI, resourceSet, feature, field.getValue());
+            newObject.eSet(feature, emfValue);
+         }
+         return newObject;
+      }
+
+      // Other EClassifiers: Enum and Datatype can't be instantiated (But they may be used to specify values with
```

Comment:
This method "creates" new elements. "creating" an enum literal doesn't make sense.

To set an enum value, we would use a "replace" operation rather than "add" (See `getEMFValue` and `getEMFEnumValue`). 

Regarding EDataTypes, there is a follow-up issue (Issue #162). It would make sense to use the createFromString() method in that context (But probably not in the `getObjectToAdd` operation; this would be a "replace" operation too. I think. There are subtle differences between "Json Patch AddOperation" and "EMF Add Command").

I guess it mostly depends on how we represent DataTypes on the json side. If they are represented a Opaque Strings, then we should use "replace". If they are specified as expanded types, then we should use "Add", and indeed add DataType support in the getObjectToAdd method (But then we wouldn't be able to use "createFromString", as we'd be receiving a structured object). Anyway, there are many special cases to consider to properly support EDataTypes, so having this discussion in the appropriate ticket (#162) makes more sense :)

Reply:
@CamilleLetavernier Thank you very much for explaining that to me. I was not aware of those fine differences between adding/creating and replacing in that context. Seeing as there is already a ticket for it, I am fine to have it like this for now since it does not prevent the rest from working as expected.